Modules/CaPTkCommon/src/CaPTkROIConstruction.cpp

```diff
@@ -2,159 +2,116 @@
-        roiConstructor.SetLatticeWindowSize(window);
-        roiConstructor.SetBoundaryCondition(fluxNeumannCondition);
-        roiConstructor.SetPatchConstructionConditionROI(patchConstructionROI);
-        roiConstructor.SetPatchConstructionConditionNone(patchConstructionNone);
-        roiConstructor.Update();
-
-        m_ROIIndices->dimension = 3;
-        m_ROIIndices->ind3D = roiConstructor.GetOutput();        
-    }
-    
-
+    // ---- Save template image for GetNext() use
+    m_MaskTemplate = mitk::Image::New();
+    m_MaskTemplate->Initialize(
+        mitk::ConvertLabelSetImageToImage(input.GetPointer())
+    ); // This creates an empty copy, with the same meta-data
+
+    m_MaskLabelSetCopy = input->GetActiveLabelSet()->Clone();
+
+    // Create helper with the mask, by calling CreateHelper
+    // AccessByItk calls the function with the ITK image (not MITK)
+    // using the correct template
+    AccessByItk(input, CreateHelper);
+
+    // TODO: Pass labels and values to helper
+    std::vector<std::string> labels;
+    std::vector<int> values;
+    this->GetLabelsAndValuesVectors(input, labels, values);
+    m_Helper->SetValuesAndNames(values, labels);
+
+    // Construct the ROI indices
+    m_Helper->Update(
+        radius,
+        step);
 }
 
-
-bool captk::ROIConstruction::HasNext()
+bool captk::ROIConstruction::IsAtEnd()
 {
-    // TODO
-    return false;
+    return m_Helper->IsAtEnd();
 }
 
-
-mitk::LabelSetImage::Pointer captk::ROIConstruction::GetNext()
+float captk::ROIConstruction::PopulateMask(mitk::LabelSetImage::Pointer& rMask)
 {
-    // TODO
-    return mitk::LabelSetImage::New();
+    rMask->Initialize(m_MaskTemplate); // This creates an empty copy, with the same meta-data
+
+    // auto label = mitk::Label::New();
+    // // auto layer = mitk::LabelSet::New();
+    // auto layer = rMask->GetActiveLabelSet();
+    // label->SetName(labelName);
+    // label->SetValue(labelValue);
+    // label->SetColor(color);
+    // layer->AddLabel(label);
+    // layer->SetActiveLabel(labelValue);
+    // rMask->AddLayer(layer);
+    // // rMask->AddLabelSetToLayer(0, mitk::LabelSet::New());
+    // // rMask->GetActiveLabelSet()->AddLabel(label);
+
+    float weight = m_Helper->PopulateMask(rMask);
+
+    // Change color of label
+    auto label = rMask->GetActiveLabelSet()->GetActiveLabel();
+    label->SetColor(m_MaskLabelSetCopy->GetLabel(label->GetValue())->GetColor());
+
+    return weight;
 }
 
 void captk::ROIConstruction::GoToBegin()
 {
-    m_CurrentIndex = 0;
+    m_Helper->GoToBegin();
+}
+
+captk::ROIConstruction& captk::ROIConstruction::operator++() //suffix
+{
+    ++*m_Helper.get(); // actual operation
+    return *this;
 }
 
-std::string captk::ROIConstruction::GetCurrentName()
+captk::ROIConstruction captk::ROIConstruction::operator++(int) //postfix(calls suffix)
 {
-    // TODO
-    return "";
+    ROIConstruction tmp(*this);
+    operator++(); // call suffix
+    return tmp;
```

Comment:
What's going on here? Should this be tmp++?

Reply:
Nope. Maybe ++tmp would work, but this is fine and clear, right?

In general what happens is:
- ++ as a prefix does the increment
- ++ as a suffix calls ++ as a prefix, which does the increment

Reply:
Apparently I've been using "suffix" to describe "prefix" in the code... I will fix it

Reply:
Updated